feature_engine/prediction/base_predictor.py

```diff
@@ -0,0 +1,235 @@
+
+from typing import List, Union
+
+import pandas as pd
+from sklearn.base import BaseEstimator
+from sklearn.pipeline import Pipeline
+from sklearn.utils.validation import check_is_fitted
+
+from feature_engine.dataframe_checks import (
+    _check_contains_inf,
+    _check_contains_na,
+    _check_input_matches_training_df,
+    _is_dataframe,
+)
+from feature_engine.discretisation import (
+    EqualFrequencyDiscretiser,
+    EqualWidthDiscretiser,
+)
+from feature_engine.encoding import MeanEncoder
+from feature_engine.variable_manipulation import (
+    _check_input_parameter_variables,
+    _find_categorical_and_numerical_variables,
+)
+
+
+class BaseTargetMeanEstimator(BaseEstimator):
+    """
+    Calculates the mean target value per category or per bin of a variable or group of
+    variables. Works with numerical and categorical variables. If variables are
+    numerical, the values are first sorted into bins of equal-width or equal-frequency.
+
+    Parameters
+    ----------
+    variables: list, default=None
+        The list of input variables. If None, the estimator will use all variables as
+        input features (except datetime).
+
+    bins: int, default=5
+        If the dataset contains numerical variables, the number of bins into which
+        the values will be sorted.
+
+    strategy: str, default='equal_width'
+        Whether the bins should of equal width ('equal_width') or equal frequency
+        ('equal_frequency').
+
+    Attributes
+    ----------
+    variables_categorical_:
+        The group of categorical input variables that will be used for prediction.
+
+    variables_numerical_:
+        The group of numerical input variables that will be used for prediction.
+
+    pipeline_:
+        A Sickit-learn Pipeline with a dicretiser and/or encoder. Used to determine the
+        mean target value per category or bin, per variable.
+
+    n_features_in_:
+        The number of features in the train set used in fit.
+
+    Methods
+    -------
+    fit:
+        Learn the mean target value per category or per bin, for each variable.
+    predict:
+        Predict using the average of the target mean value across variables.
+
+    See Also
+    --------
+    feature_engine.encoding.MeanEncoder
+    feature_engine.discretisation.EqualWidthDiscretiser
+    feature_engine.discretisation.EqualFrequencyDiscretiser
+
+    References
+    ----------
+    Adapted from:
+
+    .. [1] Miller, et al. "Predicting customer behaviour: The University of Melbourne’s
+        KDD Cup report". JMLR Workshop and Conference Proceeding. KDD 2009
+        http://proceedings.mlr.press/v7/miller09/miller09.pdf
+    """
+
+    def __init__(
+        self,
+        variables: Union[None, int, str, List[Union[str, int]]] = None,
+        bins: int = 5,
+        strategy: str = "equal_width",
+    ):
+        # boolean value can be interpreted as an integer
+        if not isinstance(bins, int) or isinstance(bins, bool):
```

Comment:
bins should just take integers, not booleans, it does not matter if booleans can be interpreted as integers.

Could we change please?

Reply:
I created the variable `_false_input_params` which is comprised of the following elements.
```
_false_input_params = [
    ("salsa", "arbitrary"),
    ("33", "mean-encoder"),
    ([7], True),
    (False, "prost"),
]
```

The assertion test associated with the last pair `(False, "prost")` was failing, i.e., a ValueError was not raised. To resolve the error, I added `or isinstance(bins, bool)` to the `if` statement to ensure a ValueError was raised in case the user inputs `True` or `False` for `bins` when instantiated the class.

I was surprised by the result. If we remove `or isinstance(bins, bool)` then the code does not pass`test_raises_error_when_wrong_input_params()`.

Reply:
Makes sense!! Thank you!